src/rgw/rgw_amqp_1.cc

```diff
@@ -0,0 +1,634 @@
+	static const int AMQP_1_TRACKER_STATUS_RECEIVED = -0x3002;
+	static const int AMQP_1_TRACKER_STATUS_ACCEPTED = -0x3003;
+	static const int AMQP_1_TRACKER_STATUS_REJECTED = -0x3004;
+	static const int AMQP_1_TRACKER_STATUS_RELEASED = -0x3005;
+	static const int AMQP_1_TRACKER_STATUS_MODIFIED = -0x3006;
+
+	static const int RGW_AMQP_1_STATUS_OK = 0x0;
+
+	int status_tracker_to_rgw(int s) {
+		if(s == AMQP_1_TRACKER_STATUS_ACCEPTED) {
+			return RGW_AMQP_1_STATUS_OK;
+		}
+	}
+
+	struct reply_callback_with_tag_t {
+		uint64_t tag;
+		reply_callback_t cb;
+
+		reply_callback_with_tag_t(uint64_t _tag, reply_callback_t _cb) : tag(_tag), cb(_cb) {}
+
+		bool operator==(uint64_t rhs) {
+			return tag == rhs;
+		}
+	};
+
+	struct reply_callback_with_tracker_t {
+		proton::tracker tracker;
+		reply_callback_t cb;
+
+		reply_callback_with_tracker_t(proton::tracker _tracker, reply_callback_t
+				_cb) : tracker(_tracker), cb(_cb) { }
+
+		// test the operator==
+		bool operator==(proton::tracker rhs) {
+			return tracker == rhs;
+		}
+	};
+
+	typedef std::vector<reply_callback_with_tag_t> CallbackList;
+	typedef std::vector<reply_callback_with_tracker_t> CallbackList_tracker;
+
+	// proton::tracker states to string
+	std::string to_string(enum proton::transfer::state s) {
+		switch(s) {
+			case proton::transfer::state::NONE:
+				return "AMQP_1_TRACKER_STATUS_UNKOWN";
+			case proton::transfer::state::RECEIVED:
+				return "AMQP_1_TRACKER_STATUS_RECEIVED";
+			case proton::transfer::state::ACCEPTED:
+				return "AMQP_1_TRACKER_STATUS_ACCEPTED";
+			case proton::transfer::state::REJECTED:
+				return "AMQP_1_TRACKER_STATUS_REJECTED";
+			case proton::transfer::state::RELEASED:
+				return "AMQP_1_TRACKER_STATUS_RELEASED";
+			case proton::transfer::state::MODIFIED:
+				return "AMQP_1_TRACKER_STATUS_MODIFIED";
+		}
+	}
+
+	int reply_to_code(enum proton::transfer::state s) {
+		switch(s) {
+			case proton::transfer::state::NONE:
+				return AMQP_1_TRACKER_STATUS_UNKOWN;
+			case proton::transfer::state::RECEIVED:
+				return AMQP_1_TRACKER_STATUS_RECEIVED;
+			case proton::transfer::state::ACCEPTED:
+				return AMQP_1_TRACKER_STATUS_ACCEPTED;
+			case proton::transfer::state::REJECTED:
+				return AMQP_1_TRACKER_STATUS_REJECTED;
+			case proton::transfer::state::RELEASED:
+				return AMQP_1_TRACKER_STATUS_RELEASED;
+			case proton::transfer::state::MODIFIED:
+				return AMQP_1_TRACKER_STATUS_MODIFIED;
+		}
+	}
+
+	struct connection_t : public proton::messaging_handler{
+		friend class Manager;
+		// proton::connection proton_conn;
+		bool marked_for_deletion = false;
+		// to count how many message been sent in total
+		uint64_t delivery_tag = 1;
+		int status;
+		mutable std::atomic<int> ref_count = 0;
+		CephContext* const cct;
+		// TODO: amqp1.0 tag?
+		CallbackList_tracker callbacks;
+		std::string broker;
+		proton::sender sender;
+		// proton::connection_options options;
+		proton::work_queue* pwork_queue;
```

Comment:
why do you need the work queue explicit here?
isn't proton working with its own work queue?